lib/Dialect/ESI/runtime/cpp/include/esi/Accelerator.h

```diff
@@ -0,0 +1,57 @@
+//===- StdServices.h - ESI standard services C++ API ------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Basic ESI APIs.
+//
+// DO NOT EDIT!
+// This file is distributed as part of an ESI package. The source for this file
+// should always be modified within CIRCT.
+//
+//===----------------------------------------------------------------------===//
+
+// NOLINTNEXTLINE(llvm-header-guard)
+#ifndef ESI_ACCELERATOR_H
+#define ESI_ACCELERATOR_H
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace esi {
+class SysInfo;
+
+/// An ESI accelerator system.
+class Accelerator {
+public:
+  static std::unique_ptr<Accelerator> connect(std::string backend,
+                                              std::string connection);
```

Comment:
I think there's a bit of an abstraction aliasing/mixup issue here, in having `Accelerator::connect` be a static function that dispatches out build calls to other registered backends - backends, which themselves have an `static TBackend::connect` function with different semantics.

I'd prefer that `Accelerator` was kept as strictly a base class of Accelerators, and then backend registration is pulled out into e.g. an `AcceleratorRegistry` class. This would then also be a singleton class, allowing you to remove `static std::map<std::string, BackendCreate> backendRegistry;` from `Accelerator.cpp`. I can't stand free variables like that, and think wrapping it inside a singleton is at least a little bit more principled and a better approach to handling the lifetime of the registry.

Reply:
I've gone back and forth on the singleton design pattern through the years. It brings me back to my Java years (*shudder*). But I hear you on the connect method being in the Accelerator class.

Reply:
I don't really see the practical advantage of the singleton design pattern given that C++ has free functions and namespaces. I've moved connect into a new `registry` namespace and the register functionality into a `registry::internal` namespace. A separate header may be desirable for internal stuff but seems like overkill for just this.